### PR TITLE
Introduce `Operator.UninstallWaitForDeletion`

### PR DESCRIPTION
**Pull request by porridge**
This is needed to work around a race/deadlock in statefulset controller tickled by now-faster kudo controller manager. See the story in https://github.com/mesosphere/kudo-cassandra-operator/pull/78 for details.

Also fix invalid format strings nearby.

### CODE REVIEW

File: pkg/kudo/operator.go [hunk 1]
```diff
@@ -169,19 +174,31 @@ func (builder OperatorBuilder) Do(client client.Client) (Operator, error) {
...
+// Waits up to timeout for the instance to be deleted, and up to 10 seconds for each of OperatorVersion and Operator.
+func (operator Operator) UninstallWaitForDeletion(timeout time.Duration) error {
 	if operator.client.Kudo == nil {
 		return fmt.Errorf("operator is not initialized")
 	}
 
 	options := metav1.DeleteOptions{}
 
+	if timeout != 0 {
+		propagationPolicy := metav1.DeletePropagationForeground
```
Comment (nfnt) [inline pkg/kudo/operator.go:194]: Foreground delete propagation is tricky. It uses finalizers and this caused issues with KUDO instances when the deletion is triggered before a deployment was completed (https://github.com/kudobuilder/kudo/issues/1071). Good that this isn't used if `timeout == 0` 👍. Let's document this caveat though, because this could result in hanging CI run if a deletion is triggered in failure conditions.

File: pkg/kudo/operator.go [hunk 2]
```diff
@@ -169,19 +174,31 @@ func (builder OperatorBuilder) Do(client client.Client) (Operator, error) {
...
+	}
+
 	err := operator.client.Kudo.
 		KudoV1beta1().
 		Instances(operator.Instance.Namespace).
 		Delete(operator.Instance.Name, &options)
 	if err != nil {
 		return fmt.Errorf(
-			"failed to delete Instance %s in namespace %s: %w",
+			"failed to delete Instance %s in namespace %s: %v",
```
Comment (nfnt) [inline pkg/kudo/operator.go]: Why is this changed?
The wrapping is on purpose. As this is a library, callers should be able to unwrap the errors here if needed. And as `test-tools` is mostly a thin wrapper around the Kubernetes API, we should expect users to have access to this API as well.

Here and below.

File: pkg/kudo/operator.go [hunk 3]
```diff
@@ -205,15 +222,74 @@ func (operator Operator) Uninstall() error {
...
 			operator.Operator.Namespace,
 			err)
 	}
 
+	if timeout != 0 {
+		kudoV1beta1 := operator.client.Kudo.KudoV1beta1()
+		i := operator.Instance
+		ov := operator.OperatorVersion
+		o := operator.Operator
+		timeoutSeconds := timeout.Round(time.Second).Milliseconds() / 1000
```
Comment (nfnt) [inline pkg/kudo/operator.go]: Nit: Hmm, or just cast `timeout.Seconds()` to `int64`? Not tested :)
```suggestion
		timeoutSeconds := int64(timeout.Seconds())
```

File: pkg/kudo/operator.go [hunk 4]
```diff
@@ -205,15 +222,74 @@ func (operator Operator) Uninstall() error {
...
+
+		err := waitForDeletion(kudoV1beta1.Instances(i.Namespace), i.ObjectMeta, timeoutSeconds)
+		if err != nil {
+			return err
+		}
+
+		err = waitForDeletion(kudoV1beta1.OperatorVersions(ov.Namespace), ov.ObjectMeta, 10)
```
Comment (nfnt) [inline pkg/kudo/operator.go]: How did you determine 10 seconds here? Would that be a safe value on a heavily loaded cluster?
Let's make this a constant.

File: pkg/kudo/operator.go [hunk 5]
```diff
@@ -205,15 +222,74 @@ func (operator Operator) Uninstall() error {
...
 		return fmt.Errorf(
-			"failed to delete Operator %s in namespace %s: %w",
+			"failed to delete Operator %s in namespace %s: %v",
 			operator.Operator.Name,
 			operator.Operator.Namespace,
 			err)
 	}
 
+	if timeout != 0 {
+		kudoV1beta1 := operator.client.Kudo.KudoV1beta1()
```
Comment (nfnt) [inline pkg/kudo/operator.go]: Nit: Let's name this `kudoClient`. IMO `kudoV1beta1` is easy to mistake with `kudov1beta1` which makes this harder to read.

**Review by nfnt — CHANGES_REQUESTED** [review]
This looks great!
I have a few minor comments and we also have to keep in mind that foreground deletion can cause issues with instances that haven't finished deployment yet. Maybe we can keep using background deletion... would the watcher work with that as well or is it tied to foreground deletion?

File: pkg/kudo/operator.go (hunk 2 above)
Comment (porridge) [reply pkg/kudo/operator.go]: Oooh, my IDE complained so I thought that was a typo until I read https://blog.golang.org/go1.13-errors now.
Looks like I need to catch up with the new hotness and upgrade my Intellij :sweat_smile: 
Reverting this part...

File: pkg/kudo/operator.go (hunk 4 above)
Comment (porridge) [reply pkg/kudo/operator.go]: This was just a guess, suggestions for a better value most welcome.
Changed to a const and added a comment.

File: pkg/kudo/operator.go (hunk 3 above)
Comment (porridge) [reply pkg/kudo/operator.go]: I'd rather keep the rounding semantics: https://play.golang.org/p/aPkzRuYuxL6

File: pkg/kudo/operator.go (hunk 1 above)
Comment (porridge) [reply pkg/kudo/operator.go:194]: Comment added.

**Comment by porridge** [comment]
PTAL @nfnt

**Review by nfnt — APPROVED** [review]
LGTM, thanks!
